### ISSUE TITLE
Interceptor redirect, chain, forward does not work in after/postHandle

### ISSUE THREAD

**Issue by orubel** [grails/grails-core#9194]
When testing with a redirect in Interceptor.after(), I am noticing that no matter how I try to do a redirect, it won't allow it when I am returning false (ie do not render view).

I figured returning true would allow me to perhaps redirect/chain/forward but instead this just went straight through to a view.

Is this a bug? Shouldn't this attempt to see if a redirect is needed before a view render?

see #600 for previous issues.

**Comment by jeffbrown** [comment]
@orubel Thanks for the feedback.

Is this happening with 3.0.4?

**Comment by jeffbrown** [comment]
I have done some testing and all of the scenarios I have tested so far have been fixed by https://github.com/grails/grails-core/commit/54449f576e4badac47d91fcebf5cb29d0e2869c6.

I have added some relevant functional tests which are currently passing.

Will investigate a bit deeper but we may be in good shape for this.

Thanks for all of your help @orubel.

**Comment by orubel** [comment]
3.0.5. Rebuilt testapp to test it. Will put up later tonight.

**Comment by orubel** [comment]
5449f5 doesn't address redirects in after, does it?

**Comment by orubel** [comment]
Can you point me to test? Maybe I can see why redirects arent working in Interceptor.after() in testapp when doing chain/redirect/forward

**Comment by jeffbrown** [comment]
> 3.0.5. Rebuilt testapp to test it. Will put up later tonight.

Awesome.

**Comment by jeffbrown** [comment]
> I figured returning true would allow me to perhaps redirect/chain/forward but instead this just went straight through to a view.

Is that a mistake are did you really mean returning `true`?

**Comment by jeffbrown** [comment]
> Can you point me to test?

The tests at https://github.com/grails/grails3-functional-tests/commit/14cfecaa3c9af269f15b4d4c7bfed44cf7a91027 and https://github.com/grails/grails3-functional-tests/commit/cb7478140ff009f69435c13cb0150c838fc08fa3 appear fail with the 3.0.5.BUILD-SNAPSHOT code base before https://github.com/grails/grails-core/commit/54449f576e4badac47d91fcebf5cb29d0e2869c6 was done and they appear to pass afterwards.  There are still more scenarios to test.  Those are the only ones I have investigated so far.

**Comment by jeffbrown** [comment]
I think invoking `redirect`, `chain` or `forward` in an `after` interceptor is probably suspect.  I am not sure if there are valid use cases for that or not.  In any case, because of the change that we made for #9183, these all appear to work now.  At least all of the scenarios I have investigated that didn't work before the #9183 change, do work now.

@orubel @graemerocher What say ye?  Can we close this?

**Comment by orubel** [comment]
Give me a second to recover from Lsst nights SeaJUG meetup; I presented last night so I didn't get a chance to build anything. I'll be putting it up this morning to show the issue. 

Would still like to see your test so I can see why mine fails and your doesn't

Redirect in Interceptor has always been acceptable as this should route back to urlmapping and to preHandle/interceptor.before() perfectly fine especially since this is an InterceptorAdapter from my understanding

**Comment by orubel** [comment]
ok just pushed up the demo code ( https://github.com/orubel/testapp )

you can test with curl --verbose --request GET --header "Content-Type:
application/json" "http://localhost:8080/api_v0.1/post/show/5"

This should redirect 5 times before outputting. Each time through it will
write:

### Before
### After
model: {...}

To show one time through loop.


Owen Rubel
415-971-0976
orubel@gmail.com

**Comment by orubel** [comment]
As you can see from demo code, it never redirects.

**Comment by jeffbrown** [comment]
```
$ curl --verbose --request GET --header "Content-Type: application/json" "http://localhost:8080/api_v0.1/post/show/5"
* Hostname was NOT found in DNS cache
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> GET /api_v0.1/post/show/5 HTTP/1.1
> User-Agent: curl/7.37.1
> Host: localhost:8080
> Accept: */*
> Content-Type: application/json
> 
< HTTP/1.1 302 Found
* Server Apache-Coyote/1.1 is not blacklisted
< Server: Apache-Coyote/1.1
< X-Application-Context: application:development
< Location: http://localhost:8080/*/post/show/5
< Content-Length: 0
< Date: Wed, 19 Aug 2015 17:30:57 GMT
< 
* Connection #0 to host localhost left intact
```

**Comment by jeffbrown** [comment]
Looks like it is redirecting.

**Comment by jeffbrown** [comment]
@orubel It looks like you deleted your comment that had that curl statement in it.  I copied that directly from your note.  Is that the scenario that is problematic for you?

**Comment by jeffbrown** [comment]
> It looks like you deleted your comment that had that curl statement in it.

Never mind that.  I see it now.  The GH interface collapsed that part of the message. My mistake.

**Comment by orubel** [comment]
Didn't delete... someone went into notes below... weird github error

And even though it sends 302, It isn't actually redirecting; I've tried with URI and get same 302 status.

if you look where it is redirecting, it is redirecting to localhost:8080/*/${controller}/{$action} 

This is incorrect. It should be trying to go to the uri and NOT the antpattern.

**Comment by jeffbrown** [comment]
> it is redirecting to localhost:8080/*/${controller}/{$action}

> This is incorrect. It should be trying to go to the uri and NOT the ant pattern.

It isn't redirecting to the ant pattern.  It is redirecting to the url that you have configured in your url mapping.  I don't know why you have '*' in the mapping, but it is there.

I sent a PR which simplifies things and demonstrates that the redirect works.

I am going to assign this to @graemerocher and leave it up to him.  As far as I can tell, it works.  There is still the question of whether it should be allowed to work, but it appears to work.

**Comment by orubel** [comment]
the * is a wildcard naturally. Rather than putting in variables in
urlmapping, I solved with a wildcard in urlmappings and allow the
interceptor to 'filter' the resulting matches. :)

Now if it's making it through the urlmapping like all other results that
map to the interceptor, then it should be getting to the interceptor.before
and continuing on.

But obviously it isn't.


Owen Rubel
415-971-0976
orubel@gmail.com

On Wed, Aug 19, 2015 at 10:54 AM, Jeff Scott Brown <notifications@github.com

**Comment by jeffbrown** [comment]
> the * is a wildcard naturally

The way to represent a wildcard in your url mapping is with `$someVarName`

**Comment by orubel** [comment]
This was the problem I was having previously and when you mentioned AntPathMatching, that solved alot of my problems. That says I can use a wildcard there and I did to allow all paths through with something in the front of the controller/action and do the filtering through at the handlerinterceptor.

Saves on filtering twice effectively. And yes, it is legal according to documentation on AntPathMatching

**Comment by jeffbrown** [comment]
> it is legal according to documentation on AntPathMatching

Is `AntPathMatching` relevant in url mapping?

**Comment by jeffbrown** [comment]
https://github.com/orubel/testapp/blob/3e01e5d5e7232eb0e729405c71a7edf1a873b836/grails-app/controllers/UrlMappings.groovy#L17

**Comment by orubel** [comment]
Isnt that what it is using? It's not regex. I mean the syntax is pretty much antpathmatching

**Comment by jeffbrown** [comment]
> Isnt that what it is using? It's not regex. I mean the syntax is pretty much antpathmatching

Kind of, but not really.  See the docs at https://github.com/grails/grails-core/blob/v3.0.4/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/RegexUrlMapping.java#L57

**Comment by jeffbrown** [comment]
You can define wildcards and capture groups, but for what you are doing there is no reason to do that.

**Comment by jeffbrown** [comment]
The mailing list is a good place to discuss that.

**Comment by orubel** [comment]
Discuss what? I dont understand. I thought I was talking about a redirect bug and you are talking about a wildcard. It works and you can easily test it without the redirect in the filter

Are you saying to discuss the redirect issue in the forums... or are you saying to discuss wildcards in urlmappings in forums? Bacuse you are the one bringing up wildcards in urlmappings.

If you want me to rewrite the test without the wildcard, I can do that. Just ask me and I'll gladly do so.

**Comment by jeffbrown** [comment]
> Discuss what?

Nevermind.  No need to discuss anything.  I think we are all set.

Thanks again.

**Comment by graemerocher** [comment]
Closing in that case as there doesn't seem to be any defined issue and this issue has become long and difficult to follow as it is capturing multiple issues in one. If there are still problems raise a secondary issue and attach an example that demonstrates it.

**Comment by orubel** [comment]
Well defined issue was with redirect in interceptor. Urlmapping seems to write a dynamic URI instead of the ACTUAL uri so upon redirect, it uses the dynamic one.

We got hung up on how I wrote the urlmapping. One can't write a URLmapping with multitenancy like that because you cant use dynamic external variables in 'static mapping'. So the alternative is the wildcard and the allow Interceptor.before() to filter.

This actually is a better solution anyway as it is single filtering rather than duplicate filtering (once at urlmapping and once at interceptor).

Unfortunately, on a redirect, that dynamic mapping is what the redirect thinks the URI is. It should be the actual URI that the URLmapping mapped to and not the urlmapping itself.

Do we still need a new ticket for this or is this clearer?
